substratevm/src/com.oracle.svm.agent.core/src/com/oracle/svm/agent/core/JvmtiAgentBase.java

```diff
@@ -0,0 +1,275 @@
+
+    /**
+     * JVMTI Agent_OnLoad callback.
+     *
+     * @param vm The JNI Java VM.
+     * @param jvmti The JVMTI environment.
+     * @param callbacks A structure that allows setting the callbacks for JVMTI events.
+     * @param options Command line options passed to the agent.
+     * @return 0 on success, anything else on failure.
+     */
+    protected abstract int onLoadCallback(JNIJavaVM vm, JvmtiEnv jvmti, JvmtiEventCallbacks callbacks, String options);
+
+    /**
+     * JVMTI Agent_OnUnload callback.
+     *
+     * @param vm The JNI Java VM.
+     * @return 0 on success, anything else on failure.
+     */
+    protected abstract int onUnloadCallback(JNIJavaVM vm);
+
+    /**
+     * JVMTI VMStart event callback.
+     *
+     * @param jvmti The JVMTI environment.
+     * @param jni The JNI environment of the thread running the JVMTI callback.
+     */
+    protected abstract void onVMStartCallback(JvmtiEnv jvmti, JNIEnvironment jni);
+
+    /**
+     * JVMTI VMInit event callback.
+     *
+     * @param jvmti The JVMTI environment.
+     * @param jni The JNI environment of the thread running the JVMTI callback.
+     * @param thread The initial thread.
+     */
+    protected abstract void onVMInitCallback(JvmtiEnv jvmti, JNIEnvironment jni, @SuppressWarnings("unused") JNIObjectHandle thread);
+
+    /**
+     * JVMTI VMDeath event callback.
+     *
+     * @param jvmti The JVMTI environment.
+     * @param jni The JNi environment of the thead running the JVMTI callback.
+     */
+    protected abstract void onVMDeathCallback(JvmtiEnv jvmti, @SuppressWarnings("unused") JNIEnvironment jni);
+
+    /**
+     * Returns the JVMTI version required by the agent.
+     *
+     * See also: {@link com.oracle.svm.agent.core.jvmti.JvmtiInterface}
+     *
+     * @return The JVMTI version required by the agent.
+     */
+    protected abstract int getRequiredJvmtiVersion();
+
+    private T handles;
+
+    /**
+     * Returns the JNI handle set constructed with {@link #constructJavaHandles}. The handles are
+     * guaranteed to be created before {@link #onLoadCallback} is called.
+     *
+     * @return The JNI handle set constructed with {@link #constructJavaHandles}.
+     */
+    public T handles() {
+        return handles;
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <T extends JNIHandleSetBase, U extends JvmtiAgentBase<T>> U singleton() {
+        VMError.guarantee(singleton != null, "The singleton class must be initialized.");
+        return (U) singleton;
+    }
+
+    /**
+     * Registers the agent singleton. This method must be called only once during a native image
+     * build.
+     *
+     * @param agentSingleton The agent implementation.
+     */
+    @SuppressWarnings({"rawtypes"})
+    protected static <T extends JNIHandleSetBase> void registerAgent(JvmtiAgentBase<T> agentSingleton) {
+        singleton = agentSingleton;
```

Comment:
Wasn't there a check that the agent hasn't been set before? I think it should stay.

Reply:
I agree, will add the check.